dali/operators/generic/cast.h

```diff
@@ -17,45 +17,62 @@
 
 #include <vector>
 
-#include "dali/pipeline/operator/operator.h"
 #include "dali/core/convert.h"
+#include "dali/core/dev_buffer.h"
+#include "dali/core/tensor_shape.h"
+#include "dali/kernels/common/block_setup.h"
+#include "dali/pipeline/operator/operator.h"
 
 namespace dali {
 
+struct CastSampleDesc {
+  void *output;
+  const void *input;
+};
 
-#define CAST_ALLOWED_TYPES \
-  (bool, uint8_t, uint16_t, uint32_t, uint64_t, int8_t, int16_t, int32_t, int64_t, \
-  float16, float, double)
+#define CAST_ALLOWED_TYPES                                                                         \
+  (bool, uint8_t, uint16_t, uint32_t, uint64_t, int8_t, int16_t, int32_t, int64_t, float16, float, \
+  double)
 
 template <typename Backend>
 class Cast : public Operator<Backend> {
  public:
-  explicit inline Cast(const OpSpec &spec) :
-    Operator<Backend>(spec),
-    output_type_(spec.GetArgument<DALIDataType>("dtype"))
-    {}
+  explicit inline Cast(const OpSpec &spec)
+      : Operator<Backend>(spec), output_type_(spec.GetArgument<DALIDataType>("dtype")) {}
 
   inline ~Cast() override = default;
 
   DISABLE_COPY_MOVE_ASSIGN(Cast);
 
  protected:
+  bool CanInferOutputs() const override {
+    return true;
+  }
```

Comment:
Food for thought. Since we're reworking all the ops to be able to infer outputs, we're going to get rid of `CanInferOutputs`, right? The question is, should we rework also return type of `SetupImpl` then?

Reply:
I had a bit different idea, that is we can split shape inference and allocation request.